### PR TITLE
Staging

### PR DESCRIPTION
**Pull request by matttyb80**
Phase 2 of BlockScience's deliverable for ICF are staged, pending review by ICF team.

### CODE REVIEW

File: Math_Specification/2_Invariants.ipynb [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    "\n",
+    "The constant values of $V_0$ upon bond/burn and $I_0$ upon attestation can be calculated from the initial values of corresponding state variables. \n",
+    "\n",
+    "| | Bond/Burn | Attest Positive/Negative |\n",
+    "| :-: | :-: | :-: |\n",
+    "| $V(R, S)$ | $V = \\frac{S^\\kappa}{R}$ | ? |\n",
+    "| $I(\\alpha, R)$ | ? | $I = \\frac{C\\alpha}{1-\\rho} = R + C\\alpha$ |\n",
+    "\n",
+    "### Change in $V$ upon Attestation\n",
+    "Consider that an agent has made an postive attestation action. The changes in the state variables corresponding to the attestation are\n",
```
Comment (hxrts) [inline Math_Specification/2_Invariants.ipynb]:
```suggestion
    "Consider that an agent has made a positive attestation action. The changes in the state variables corresponding to the attestation are\n",
```

File: Math_Specification/1_System_Specification.ipynb [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "## Second Tier Mechanisms\n",
+    "### Overcollateralization\n",
+    "As $\\alpha \\rightarrow 0$, the system needs to over-reserve to protect against it being unable to pay back investors at Settlement. If the system does not exceed a pre-specified maximum and minimum collateralization, then the total payout should be invariant.\n",
```
Comment (hxrts) [inline Math_Specification/1_System_Specification.ipynb]:
```suggestion
    "As $\\alpha \\rightarrow 0$, the system needs to over-reserve to protect against it being unable to pay back investors at Settlement. If the system does not fall within the bounds of a pre-specified minimum and maximum collateralization, then the total payout should be invariant.\n",
```

File: Math_Specification/3_AlphaUpdate.ipynb [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    "$$(\\alpha C + R) \\frac{S_{free}}{S}$$ \n",
+    "\n",
+    "Aggregate for holders of the negative outcome claims $Q_0$\n",
+    "$$(1-\\alpha) \\frac{S_0}{S} R$$\n",
+    "\n",
+    "Aggregate for holders of the positive outcome claims $Q_1$\n",
+    "$$\\alpha \\frac{S_1}{S} (C+R)$$\n",
+    "\n",
+    "Note that the total expected payout for the whole system is given by\n",
+    "$$\\mathbb E(\\Theta) = (\\alpha C + R) \\frac{S_{free}}{S}\n",
```
Comment (hxrts) [inline Math_Specification/3_AlphaUpdate.ipynb]: not rendering correctly

File: Math_Specification/3_AlphaUpdate.ipynb [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    "$S$: total bond token supply <br/>\n",
+    "$S_0$: total bond tokens locked in negative attestation <br/>\n",
+    "$S_1$: total bond tokens locked in positive attestation <br/>\n",
+    "$S_{free} = S-S_0-S_1$: bond tokens not committed to attestations <br/>\n",
+    "$Q_0$: total claims on rewards given negative attestation <br/>\n",
+    "$Q_1$: total claims on rewards given  positive attestation <br/>\n",
+    "$\\alpha$: the systems current belief of the liklihood of success\n",
+    "\n",
+    "### Expected Payout\n",
+    "The close-out 'expected' payout at the system level can be described as <br/>\n",
```
Comment (hxrts) [inline Math_Specification/3_AlphaUpdate.ipynb]: I think this line could be a bit more descriptive to setup the expressions below.

File: Math_Specification/3_AlphaUpdate.ipynb [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    "\n",
+    "If $S_0=S_1$:\n",
+    "\n",
+    "Then the above collapses to $(\\alpha C + R)$.\n",
+    "\n",
+    "Now let's suppose $S_0\\not=S_1$ but let's find $\\alpha$ such that this equality holds anyway:\n",
+    "\n",
+    "$(\\alpha C + R) = (\\alpha C + R) \\frac{S_{free}}{S} + \\alpha \\frac{S_1}{S}C + \\frac{S_0}{S}R + \\alpha R \\frac{S_1-S_0}{S}$\n",
+    "\n",
+    "Distributing the multiplications:\n",
```
Comment (hxrts) [inline Math_Specification/3_AlphaUpdate.ipynb]: never seen "multiplications" used like this

File: Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb [hunk 6]
```diff
@@ -0,0 +1,876 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Step 1: Draw $\\hat P$ from a signal representing movement of $\\hat P$ over time**\n",
+    "\n",
+    "Since the agent's private belief of price $\\hat P$ is unknown to the designer, it can be sampled from a signal that represents its value at a given timestep. \n",
+    "\n",
+    "The designer may test the system using the following $\\hat P$ candidate signals:\n",
```
Comment (hxrts) [inline Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb:35]: "candidate signals" could use a little more context, both what you mean in general and each specific signal

File: Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb [hunk 7]
```diff
@@ -0,0 +1,876 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Step 4: Determine direction and size of agent's action**\n",
+    "\n",
+    "\n",
+    "Let $\\tau$ be the threshold condition for the agent to make a bond/burn action. $P$ and $\\hat P$ need to differ by an amount greater than $\\tau$ for the agent to make a bond/burn action. \n",
+    "\n",
+    "We computed a temporary $max_{amt}$ in Step 3 using Armijo rule. <br/>\n",
```
Comment (hxrts) [inline Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb]:
```suggestion
    "We computed a temporary $max_{amt}$ in Step 3 using the Armijo rule. <br/>\n",
```

File: Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb [hunk 8]
```diff
@@ -0,0 +1,876 @@
...
+    "\n",
+    "If $\\bar P - \\hat P > \\tau$, the agent *burns* with an $amttoburn = amt * \\beta$\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "1. Compare Phat to P w/ tau to determine direction, employ hystersis\n",
```
Comment (hxrts) [inline Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb]: use symbols to be consistent, don't abbreviate "with"

File: Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb [hunk 9]
```diff
@@ -0,0 +1,876 @@
...
+    "    - Bond \n",
+    "    - do nothing\n",
```
Comment (hxrts) [inline Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb]:
```suggestion
    "    - Do nothing\n",
```

File: Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb [hunk 10]
```diff
@@ -0,0 +1,876 @@
...
+    "    - Burn\n",
+    "2. If bond\n",
+    "    deltaR = max_action = r(1-dust)\n",
```
Comment (hxrts) [inline Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb]: again, try to be consistent with use of symbols here

File: Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb [hunk 11]
```diff
@@ -0,0 +1,876 @@
...
+    "ax.set(xlabel='Beta(0.1,0.1)', ylabel='Frequency')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 48,
+   "metadata": {},
+   "outputs": [
+    {
+     "ename": "AttributeError",
```
Comment (hxrts) [inline Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb]: error here

File: Math_Specification/PriceInitialization_and_OutcomeIndifference.ipynb [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    "\n",
+    "Since outcome payout is conserved throughout the duration of the bond, we have $\\frac{\\Theta_{0}}{S_0} =\\frac{\\Theta}{S}$. Rearranging, $$\\frac{\\Theta_{0}}{S_0}=\\frac{\\Theta}{S}$$<br> For this equivalency to hold, in the definition for $\\Theta_i$, we require that $$\\frac{1}{\\rho}\\frac{R}{\\Theta} = 1$$<br> Rearranging to obtain $R$, $$R=\\rho\\Theta$$<br> Since $\\Theta_{true}$ is unknown, the system and its agents need to estimate it. Plugging the above in $\\Theta_{true}$, we have $$\\Theta_{true}=C\\Omega + \\rho\\Theta_{true}$$ Simplifying, $$\\Theta_{true}=\\frac{C\\Omega}{1-\\rho}$$<br> From here, we derive the system's estimate of total payout and the agents' estimate of individual payout.\n",
+    "\n",
+    "#### 3. System's estimate of Total Payout\n",
+    "From (1) we have $$\\mathbb{E}(\\Theta_{sys}) = C\\alpha + R$$<br> Due to the analogy between $\\Theta_{sys}$ and $\\Theta_{true}$, we can conclude $$\\mathbb{E}(\\Theta_{sys})= \\frac{C\\alpha}{1-\\rho}$$<br>\n",
```
Comment (hxrts) [inline Math_Specification/PriceInitialization_and_OutcomeIndifference.ipynb]: could you be a little more explicit about what you mean about the analogy here

File: Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb [hunk 13]
```diff
@@ -0,0 +1,876 @@
...
+    "This attestation action in turn moves the system alpha $\\alpha$ closer to the agent's private belief of alpha, $\\hat\\alpha$.\n",
+    "\n",
+    "The system alpha state variable changes from $\\alpha_{t} \\rightarrow \\alpha_{t+1}$, computed by\n",
+    "\n",
+    "$$\\alpha_{t+1} = (\\epsilon) \\alpha + (1-\\epsilon)\\cdot(1-\\delta) \\alpha + (1-\\epsilon).(\\delta) \\bar\\alpha$$\n",
+    "<br/> where $\\epsilon$ is the designer-defined fixed weight between [0,1] and $\\delta=\\frac{\\Delta s}{S_1+S_0+\\Delta s}$ is the dynamic weight.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (hxrts) [inline Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb:195]: Each of the plots below could use a consistent title and description of what's being plotted. A couple have a small blurb, but more regularity in the formatting and descriptions that setup why these were plotted would be very helpful.

File: Math_Specification/4b_AgentBehaviourModel - Analytic.ipynb [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    "\n",
+    "Substitute $\\Delta q_1 = Q_1 \\sqrt{(1+\\frac{\\Delta s_{min}}{S}) - 1}$. Solve for $\\Delta s_{min}$\n",
+    "\n",
+    "$$\\Delta s_{min} = 0$$\n",
+    "\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (hxrts) [inline Math_Specification/4b_AgentBehaviourModel - Analytic.ipynb:104]: I assume this is a placeholder for the next milestone.

**Review by hxrts — APPROVED** [review]
Hi all, this is great work! Thanks for your patience, just a few small requests.

File: Math_Specification/3_AlphaUpdate.ipynb (hunk 3 above)
Comment (matttyb80) [reply Math_Specification/3_AlphaUpdate.ipynb]: removed tabs

File: Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb (hunk 11 above)
Comment (matttyb80) [reply Math_Specification/4a_AgentBehaviourModel - Heuristic.ipynb]: subsequent plot replaces this one

**Review by Yedfree — APPROVED** [review]

**Pull request merged**